multicluster/controllers/multicluster/member/clusterset_controller.go

```diff
@@ -90,57 +93,71 @@ func NewMemberClusterSetReconciler(client client.Client,
+				// when its MemberClusterAnnounce is deleted.
+				if err := r.deleteMemberClusterAnnounce(ctx); err != nil {
+					// MemberClusterAnnounce could be kept in the leader cluster, if antrea-mc-controller crashes after the failure.
+					// Leader cluster will delete the stale MemberClusterAnnounce with a garbage collection mechanism in this case.
+					return fmt.Errorf("failed to delete MemberClusterAnnounce in the leader cluster: %v", err)
+				}
+				r.remoteCommonArea.Stop()
+				r.remoteCommonArea = nil
+				r.clusterSetConfig = nil
+				r.clusterID = common.InvalidClusterID
+				r.clusterSetID = common.InvalidClusterSetID
+			}
+			klog.InfoS("Clean up all resources created by Antrea Multi-cluster Controller", "clusterset", req.NamespacedName)
+			if err := cleanUpResourcesCreatedByMC(ctx, r.Client); err != nil {
+				return err
+			}
+			return nil
 		}
-		return ctrl.Result{}, nil
-	}
 
-	klog.InfoS("Received ClusterSet add/update", "clusterset", klog.KObj(clusterSet))
+		klog.InfoS("Received ClusterSet add/update", "clusterset", klog.KObj(clusterSet))
 
-	// Handle create or update
-	if r.clusterSetConfig == nil {
-		r.clusterID, err = common.GetClusterID(r.ClusterCalimCRDAvailable, req, r.Client, clusterSet)
-		if err != nil {
-			return ctrl.Result{}, err
-		}
-		r.clusterSetID = common.ClusterSetID(clusterSet.Name)
-		if clusterSet.Spec.ClusterID == "" {
-			// ClusterID is a required feild, and the empty value case should only happen
-			// when Antrea Multi-cluster is upgraded from an old version prior to v1.13.
-			// Here we try to update the ClusterSet's ClusterID when it's configured in an
-			// existing ClusterClaim.
-			clusterSet.Spec.ClusterID = string(r.clusterID)
-			err = r.Update(context.TODO(), clusterSet)
+		// Handle create or update
+		if r.clusterSetConfig == nil {
+			r.clusterID, err = common.GetClusterID(r.ClusterCalimCRDAvailable, req, r.Client, clusterSet)
 			if err != nil {
-				klog.ErrorS(err, "Failed to update ClusterSet's ClusterID", "clusterset", req.NamespacedName)
-				return ctrl.Result{}, err
+				return err
+			}
+			r.clusterSetID = common.ClusterSetID(clusterSet.Name)
+			if clusterSet.Spec.ClusterID == "" {
+				// ClusterID is a required feild, and the empty value case should only happen
+				// when Antrea Multi-cluster is upgraded from an old version prior to v1.13.
+				// Here we try to update the ClusterSet's ClusterID when it's configured in an
+				// existing ClusterClaim.
+				clusterSet.Spec.ClusterID = string(r.clusterID)
+				err = r.Update(context.TODO(), clusterSet)
+				if err != nil {
+					klog.ErrorS(err, "Failed to update ClusterSet's ClusterID", "clusterset", req.NamespacedName)
+					return err
+				}
 			}
 		}
-	}
-	r.clusterSetConfig = clusterSet.DeepCopy()
+		r.clusterSetConfig = clusterSet.DeepCopy()
 
-	err = r.createOrUpdateRemoteCommonArea(clusterSet)
-	if err != nil {
+		err = r.createOrUpdateRemoteCommonArea(clusterSet)
```

Comment:
It will check ClusterSet changes and stop commonArea for update case. I feel it's better to defer this kind of change, I found that ClusterSet controller can't handle update case very well in existing implementation. I can refine it and do clean up all together in later PR but need more tests and verification. if we introduce the related changes now, this change might not be ready in v1.14.

Reply:
Check my last comment.